File: 199.二叉树的右视图.py
#
# @lc app=leetcode.cn id=199 lang=python3
#
# [199] 二叉树的右视图
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from queue import Queue


class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        right_side_view = {}
        q = Queue(maxsize=0)
        q.put([root, 0])
        max_depth = -1
        while not q.empty():
            node, depth = q.get()
            if node:
                max_depth = max(max_depth, depth)
                right_side_view[depth] = node.val
                q.put([node.left, depth+1])
                q.put([node.right, depth+1])
        return [right_side_view[depth] for depth in range(max_depth+1)]
# @lc code=end

File: 199.二叉树的右视图.cpp
/*
 * @lc app=leetcode.cn id=199 lang=cpp
 *
 * [199] 二叉树的右视图
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        unordered_map<int, int> right_side_view;
        int max_depth = -1;
        queue<TreeNode *> nodes;
        queue<int> depths;
        nodes.push(root);
        depths.push(0);
        while (!nodes.empty())
        {
            TreeNode *node = nodes.front();
            nodes.pop();
            int depth = depths.front();
            depths.pop();
            if (node)
            {
                max_depth = max(max_depth, depth);
                right_side_view[depth] = node->val;
                nodes.push(node->left);
                nodes.push(node->right);
                depths.push(depth + 1);
                depths.push(depth + 1);
            }
        }
        vector<int> result;
        for (int i = 0; i <= max_depth; i++)
        {
            result.push_back(right_side_view[i]);
        }
        return result;
    }
};
// @lc code=end
